### PR TITLE
fix: filtering unsupported codefixes in glint-fix plugin

### LINKED ISSUE #1184: Unsafe codefix applied in Glint `.gts` file for arrow function
A suggestion is being applied in a `.gts` file that breaks syntax.

**Input**

```
export default class MyForm {
  submit(e: Event) {
    const something = {
      run: function () {
        console.log(this);
      }.bind(this),
    };

    return something;
  }
}
```

**Output**
```
// @ts-expect-error @rehearsal TODO TS2345: Argument of type 'typeof MyForm' is not assignable to parameter of type 'abstract new () => HasContext<AnyContext'.
export default class MyForm {
  submit(e: Event) {
    const something = {
      run:  () => {
        console.log(this);
      }.bind(this),
    };

    return something;
  }
}
```

Screenshot 2023-07-21 at 3 14 48 PM

### PR DESCRIPTION
**Pull request by egorio**
Create a glint-codefixes-collection (extends typescript-codefixes-collection) to incapsulate the logic to getting and filtering out unsupported codefixes. 

Glint service is using VSCode CodeAction and doesn't provide codefixes' fixName value. So glint-codefixes-collection is using new filtering function based on fixDescription values.

Also fixes and improves:
- supported diagnostic codes (description codes was used by mistake)
- comparison of text changes which affected type inference of function params
- increase limit value in drain mode

Fixes #1184

### CODE REVIEW

File: packages/smoke-test/test/commands/fix.test.ts [hunk 1]
```diff
@@ -47,7 +47,7 @@ describe('fix command', () => {
     expectFile(join(projectRoot, 'src/app.ts')).toMatchSnapshot();
   });
 
-  test('ember_js_app_4.11', async () => {
+  test('ember_js_app_4.11 --mode=single-pass', async () => {
```
Comment (egorio) [inline packages/smoke-test/test/commands/fix.test.ts:50]: The `drain` is default mode now, need to test opposite - `single-pass`.

File: packages/codefixes/src/codefixes.ts [hunk 2]
```diff
@@ -25,7 +26,11 @@ export const codefixes = new CodeFixesProvider([
 ]);
 
 export const glintCodeFixes = new CodeFixesProvider([
-  new BaseCodeFixCollection([new AddMissingArgToComponentSignature()]),
+  new GlintCodeFixCollection(),
+  new BaseCodeFixCollection([
+    // Need to be run after standard "typedef to type" fix applied
+    new AddMissingArgToComponentSignature(),
```
Comment (wondersloth) [inline packages/codefixes/src/codefixes.ts:32]: Why would this not be in GlintCodeFixCollection? I understand it needs to be last.

File: packages/migrate/test/__snapshots__/migrate.test.ts.snap [hunk 3]
```diff
@@ -245,6 +245,28 @@ export default class Hello extends Component<HelloSignature> {
...
 `;
 
+exports[`fix > .gts > with more errors 1`] = `
+"// @ts-expect-error @rehearsal TODO TS2345: Argument of type 'typeof MyForm' is not assignable to parameter of type 'abstract new () => HasContext<AnyContext'.
+export default class MyForm {
+  submit(e: Event) {
+    const something = {
+      run:  () => {
+        console.log(this);
+      }.bind(this),
```
Comment (wondersloth) [inline packages/migrate/test/__snapshots__/migrate.test.ts.snap:255]: This is invalid syntax.
https://www.typescriptlang.org/play?#code/MYGwhgzhAECyCeBhA9gWwA7IHYFMsBcAoAb2kOggFcAjVAS3wAocAuaAUQDc98BKaYuWjDg2CPgpoc+ABZ0sAc2gBeAUOHCATpSxtojfsoB8ajWeiisEZCBwA6EMgWNZdCLwDc6jQF871eQATFzl3ABpvHy8zTWlKTSxJVGk5RWjoH0JMoA
Screenshot 2023-07-26 at 4 16 30 PM

File: packages/plugins/src/plugins/diagnostic-fix.plugin.ts [hunk 4]
```diff
@@ -78,7 +78,7 @@ export class DiagnosticFixPlugin extends Plugin<DiagnosticFixPluginOptions> {
     // In the drain mode diagnostics list is getting refreshed in every cycle which might have end up
     // with more error need to be fixed then was originally. The limit based on original amount of diagnostics
     // helps to avoid an infinitive loop in some edge cases when new errors keep coming when previous fixed.
-    let limit = diagnostics.length * 2;
+    let limit = diagnostics.length * 10;
```
Comment (wondersloth) [inline packages/plugins/src/plugins/diagnostic-fix.plugin.ts:81]: Thank you for bumping this.

File: packages/smoke-test/test/commands/__snapshots__/fix.test.ts.snap [hunk 5]
```diff
@@ -443,7 +443,7 @@ stdout.write('\\\\x1b[?25l');
 "
 `;
 
-exports[`fix command > ember_js_app_4.11 --mode=drain 1`] = `
```
Comment (wondersloth) [inline packages/smoke-test/test/commands/__snapshots__/fix.test.ts.snap:446]: Why is the drain test missing? Did it get moved?

File: packages/smoke-test/test/commands/fix.test.ts (hunk 1 above)
Comment (wondersloth) [reply packages/smoke-test/test/commands/fix.test.ts:50]: 🤣 My question above is answered by this. 👍

**Review by wondersloth — CHANGES_REQUESTED** [review]
Some minor comments, more curious about the `() => {}.bind(this)` not being suppressed.

File: packages/codefixes/src/codefixes.ts (hunk 2 above)
Comment (egorio) [reply packages/codefixes/src/codefixes.ts:32]: Glint collection provides codefixes provided by Glin service (typescript ones + glint ones). Base collection is for our own.

File: packages/migrate/test/__snapshots__/migrate.test.ts.snap (hunk 3 above)
Comment (egorio) [reply packages/migrate/test/__snapshots__/migrate.test.ts.snap:255]: WIP. I allowed this codefix to be applied just for regex test during development. Will be disabled soon.

File: packages/codefixes/Add-Fixes.md
[the remark this answers is not in the document]
Comment (egorio) [reply packages/codefixes/Add-Fixes.md:1]: Thanks!

File: packages/migrate/test/__snapshots__/migrate.test.ts.snap (hunk 3 above)
Comment (egorio) [reply packages/migrate/test/__snapshots__/migrate.test.ts.snap]: Disabled now.

File: packages/codefixes/src/codefixesMessages.json [hunk 6]
```diff
@@ -2,170 +2,209 @@
...
-      2356, 2362, 2363, 2736, 2365, 2367, 2801, 2461, 2495, 2802, 2549, 2548, 2488, 2504, 2345,
-      2339, 2349, 2351
+      2356, 2362, 2363, 2736, 2365, 2367, 2801, 2461, 2495, 2802, 2549, 2548, 2488, 2504, 2345, 2339, 2349, 2351
     ]
   },
   "addMissingAwaitToInitializer": {
-    "title": "Add `await` To Intializers",
+    "title": "Add `await` To Initializers",
     "description": "Adds missing `await` keyword to declarations and expressions.",
-    "diagnostics": [95084, 95089]
```
Comment (egorio) [inline packages/codefixes/src/codefixesMessages.json:18]: Fixed: codefixes description was used as error codes.

File: packages/ts-utils/src/helpers/strings.ts [hunk 7]
```diff
@@ -24,9 +24,12 @@ export function applyTextChanges(text: string, textChanges: TextChange[]): strin
   return text;
 }
 
+/**
+ * Compares two text changes
+ */
 export function isSameChange(a: TextChange, b: TextChange): boolean {
   return (
-    a.span.start === a.span.start && a.span.length === b.span.length && a.newText === b.newText
+    a.span.start === b.span.start && a.span.length === b.span.length && a.newText === b.newText
```
Comment (egorio) [inline packages/ts-utils/src/helpers/strings.ts:32]: Fixed: only the first text change on the line was allied (ex. function params' types inference)

**Review by lynchbomb — APPROVED** [review]

**Pull request merged**